### PR TITLE
NF+BF: _verified_reverse1() to add safety net to Mapper.reverse1()

### PR DESCRIPTION
**Pull request by hanke**
Minimally invasive patch to protect all critical calls to reverse1() in the code base. All remaining occurrences are in the docs, or in places where failure should be obvious.

### CODE REVIEW

File: mvpa2/mappers/base.py [hunk 1]
```diff
@@ -32,6 +32,34 @@ def _assure_consistent_a(ds, oshape):
...
+
+    Returns
+    -------
+    array
+      Shape matches a single sample in terms of the mappers input space.
+    """
+    dummy_axis_sample = np.asanyarray(onesample)[None]
+    rsample = mapper.reverse(dummy_axis_sample)
+    if not len(rsample) == 1:
+        warning("Reverse mapping single sample yielded multiple -- can lead to uninteded behavior!")
```
Comment (yarikoptic) [inline mvpa2/mappers/base.py]: Typo unintended

**Comment by yarikoptic** [comment]
Not sure yet how it helps if mapper simply doesn't do anything - does it still fail that _2  of mine with a warning?

**Comment by hanke** [comment]
It does not help in that regard, it simply tests whether reverse1 is
actually just one.

On Wed, Jan 6, 2016 at 3:21 PM, Yaroslav Halchenko <notifications@github.com
> wrote:

> Not sure yet how it helps if mapper simply doesn't do anything - does it
> still fail that _2 of mine with a warning?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PyMVPA/PyMVPA/pull/394#issuecomment-169334257>.
>



-- 
Michael Hanke
http://mih.voxindeserto.de

**Comment by hanke** [comment]
I merged your PR into this one to avoid going nuts. _1 is now updated with _verified_reverse1 to show that is works/helps.

**Comment by hanke** [comment]
BTW: Docs of `get_mapped()` clearly state the requirement of a _trained_ mapper.

**Comment by hanke** [comment]
Pushed again with an update of the second test. Works like charm. Merge and close ;-)

File: mvpa2/tests/test_mapper.py [hunk 2]
```diff
@@ -470,3 +476,49 @@ def test_identity_mapper(s):
...
+    ds = ds_orig.get_mapped(m)
+    ds0_rev = ds.a.mapper.reverse1(ds.samples[0])  # should work
+    assert_equal(ds0_rev.shape, (ds_orig.nfeatures,))
+
+    # direct feature selection
+    ds_ = ds[:, [0, 2]]
+    # should work but doesn't due to
+    # RuntimeError: Cannot reverse-map data since the original data shape is unknown. Either set `dshape` in the constructor, or call train().
+    #ds0_rev_ = ds_.a.mapper.reverse1(ds_.samples[0])
+    ds0_rev_ = _verified_reverse1(ds_.a.mapper, ds_.samples[0])
```
Comment (yarikoptic) [inline mvpa2/tests/test_mapper.py]: so this is a workaround or above direct  commented reverse1() out should have worked too? ;)

File: mvpa2/tests/test_mapper.py [hunk 3]
```diff
@@ -470,3 +476,49 @@ def test_identity_mapper(s):
...
+    ds_orig_rev0 = ds_orig.a.mapper.reverse1(ds_orig.samples[0])
+    # would have gone all the way to 3d
+    assert_equal(ds_orig_rev0.ndim, 3)
+
+    bool_mask = np.ones((ds_orig.nfeatures,), dtype=bool)
+    bool_mask[0] = False
+
+    for idx in (bool_mask, [0, 2]):
+        ds = ds_orig[:, idx]
+        ds_rev0 = _verified_reverse1(ds.a.mapper, ds.samples[0])
```
Comment (yarikoptic) [inline mvpa2/tests/test_mapper.py]: same here -- reverse1 - should have worked or not?

**Comment by yarikoptic** [comment]
NB test failure above is unrelated precision issue

**Comment by hanke** [comment]
No, reverse1() would not, and did not work. It uses a different (flawed)
logic.

On Wed, Jan 6, 2016 at 5:22 PM, Yaroslav Halchenko <notifications@github.com
> wrote:

> NB test failure above is unrelated precision issue
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PyMVPA/PyMVPA/pull/394#issuecomment-169376872>.
>



-- 
Michael Hanke
http://mih.voxindeserto.de

**Comment by yarikoptic** [comment]
so shouldn't we do something about it???

**Comment by hanke** [comment]
Kill reverse1? I am all for it. But maybe not right now. Or maybe right
now...

**Pull request merged**